File: skills/systematic-debugging/condition-based-waiting-example.rb
# Complete implementation of condition-based waiting utilities
# Adapted from: test infrastructure improvements (2025-10-03)
# Context: Fixed 15 flaky tests by replacing arbitrary sleeps

require "timeout"

module ConditionWaiting
  # Wait for a condition block to return a truthy value.
  #
  # @param description [String] human-readable label for timeout errors
  # @param timeout [Numeric] seconds before raising (default 5)
  # @param interval [Numeric] polling interval in seconds (default 0.01)
  # @yield condition to evaluate; return truthy to stop waiting
  # @return the truthy value returned by the block
  # @raise [RuntimeError] if timeout expires
  def wait_for(description: "condition", timeout: 5, interval: 0.01)
    deadline = Process.clock_gettime(Process::CLOCK_MONOTONIC) + timeout
    loop do
      result = yield
      return result if result
      if Process.clock_gettime(Process::CLOCK_MONOTONIC) > deadline
        raise "Timeout after #{timeout}s waiting for: #{description}"
      end
      sleep interval
    end
  end

  # Wait for a specific event type to appear in a thread or queue.
  #
  # @param source [#events] object that responds to #events returning an array
  # @param event_type [Symbol, String] event type to wait for
  # @param timeout [Numeric] seconds before raising
  # @return [Hash] the matching event
  def wait_for_event(source, event_type, timeout: 5)
    wait_for(description: "event :#{event_type}", timeout: timeout) do
      source.events.find { |e| e[:type] == event_type }
    end
  end

  # Wait until at least +count+ events of +event_type+ have been emitted.
  #
  # @param source [#events] object that responds to #events
  # @param event_type [Symbol, String] event type to count
  # @param count [Integer] minimum required count
  # @param timeout [Numeric] seconds before raising
  # @return [Array<Hash>] all matching events once count is reached
  def wait_for_event_count(source, event_type, count:, timeout: 5)
    wait_for(
      description: "#{count}x event :#{event_type}",
      timeout: timeout
    ) do
      matches = source.events.select { |e| e[:type] == event_type }
      matches if matches.length >= count
    end
  end

  # Wait for an event matching both type and predicate.
  #
  # @param source [#events] object that responds to #events
  # @param event_type [Symbol, String] event type to match
  # @param timeout [Numeric] seconds before raising
  # @yield [event] additional predicate; return truthy to accept the event
  # @return [Hash] the first matching event
  def wait_for_event_match(source, event_type, timeout: 5, &predicate)
    wait_for(
      description: "event :#{event_type} matching predicate",
      timeout: timeout
    ) do
      source.events.find { |e| e[:type] == event_type && predicate.call(e) }
    end
  end
end

File: skills/systematic-debugging/condition-based-waiting-example.dart
// Complete implementation of condition-based waiting utilities.
// Adapted from: test infrastructure improvements (2025-10-03).
// Context: Fixed 15 flaky tests by replacing arbitrary delays.

import 'dart:async';

/// A single recorded event, e.g. emitted by a fake repository, a mocktail
/// spy, or a bloc's recorded state transitions.
class DebugEvent {
  const DebugEvent(this.type, [this.data]);

  final String type;
  final Object? data;
}

/// Anything that exposes a growing log of [DebugEvent]s to poll against.
abstract class EventSource {
  List<DebugEvent> get events;
}

/// Waits for [condition] to return a non-null value, polling every
/// [interval] until [timeout] elapses.
///
/// Returns the truthy value produced by [condition].
/// Throws a [TimeoutException] labelled with [description] if the
/// deadline passes first.
Future<T> waitFor<T>(
  FutureOr<T?> Function() condition, {
  String description = 'condition',
  Duration timeout = const Duration(seconds: 5),
  Duration interval = const Duration(milliseconds: 10),
}) async {
  final deadline = DateTime.now().add(timeout);
  while (true) {
    final result = await condition();
    if (result != null) return result;
    if (DateTime.now().isAfter(deadline)) {
      throw TimeoutException(
        'Timeout after ${timeout.inSeconds}s waiting for: $description',
      );
    }
    await Future<void>.delayed(interval);
  }
}

DebugEvent? _firstEventOfType(List<DebugEvent> events, String eventType) {
  for (final event in events) {
    if (event.type == eventType) return event;
  }
  return null;
}

/// Waits for a specific event type to appear in [source].
Future<DebugEvent> waitForEvent(
  EventSource source,
  String eventType, {
  Duration timeout = const Duration(seconds: 5),
}) {
  return waitFor(
    () => _firstEventOfType(source.events, eventType),
    description: 'event $eventType',
    timeout: timeout,
  );
}

/// Waits until at least [count] events of [eventType] have been emitted.
///
/// Returns all matching events once the count is reached.
Future<List<DebugEvent>> waitForEventCount(
  EventSource source,
  String eventType, {
  required int count,
  Duration timeout = const Duration(seconds: 5),
}) {
  return waitFor(
    () {
      final matches =
          source.events.where((e) => e.type == eventType).toList();
      return matches.length >= count ? matches : null;
    },
    description: '${count}x event $eventType',
    timeout: timeout,
  );
}

/// Waits for an event matching both [eventType] and [predicate].
Future<DebugEvent> waitForEventMatch(
  EventSource source,
  String eventType,
  bool Function(DebugEvent event) predicate, {
  Duration timeout = const Duration(seconds: 5),
}) {
  return waitFor(
    () {
      for (final event in source.events) {
        if (event.type == eventType && predicate(event)) return event;
      }
      return null;
    },
    description: 'event $eventType matching predicate',
    timeout: timeout,
  );
}
